deeplearning4j/deeplearning4j-nlp-parent/deeplearning4j-nlp/src/main/java/org/deeplearning4j/iterator/BertIterator.java

```diff
@@ -0,0 +1,479 @@
+import org.deeplearning4j.text.tokenization.tokenizer.Tokenizer;
+import org.deeplearning4j.text.tokenization.tokenizerfactory.BertWordPieceTokenizerFactory;
+import org.deeplearning4j.text.tokenization.tokenizerfactory.TokenizerFactory;
+import org.nd4j.base.Preconditions;
+import org.nd4j.linalg.api.buffer.DataType;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.dataset.api.MultiDataSet;
+import org.nd4j.linalg.dataset.api.MultiDataSetPreProcessor;
+import org.nd4j.linalg.dataset.api.iterator.MultiDataSetIterator;
+import org.nd4j.linalg.factory.Nd4j;
+import org.nd4j.linalg.indexing.NDArrayIndex;
+import org.nd4j.linalg.primitives.Pair;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * BertIterator is a MultiDataSetIterator for training BERT (Transformer) models in the following way:<br>
+ * (a) Unsupervised - Masked language model task (no sentence matching task is implemented thus far)<br>
+ * (b) Supervised - For sequence classification (i.e., 1 label per sequence, typically used for fine tuning)<br>
+ * The task can be specified using {@link Task}.
+ * <br>
+ * Example for unsupervised training:<br>
+ * <pre>
+ * {@code
+ *
+ * }
+ * </pre>
+ *
+ * This iterator supports numerous ways of configuring the behaviour with respect to the sequence lengths and data layout.<br>
+ * <br>
+ * <u><b>{@link LengthHandling} configuration:</b></u><br>
+ * Determines how to handle variable-length sequence situations.<br>
+ * <b>FIXED_LENGTH</b>: Always trim longer sequences to the specified length, and always pad shorter sequences to the specified length.<br>
+ * <b>ANY_LENGTH</b>: Output length is determined by the length of the longest sequence in the minibatch<br>
+ * <b>CLIP_ONLY</b>: For any sequences longer than the specified maximum, clip them. If the maximum sequence length in
+ * a minibatch is shorter than the specified maximum, no padding will occur.<br>
+ *<br><br>
+ * <u><b>{@link FeatureArrays} configuration:</b></u><br>
+ * Determines what arrays should be included.
+ * <b>INDICES_MASK</b>: Indices array and mask array only, no segment ID array<br>
+ * <b>INDICES_MASK</b>: Indices array, mask array and segment ID array (which is all 0s for single segment tasks)<br>
+ * <br>
+ * <u><b>{@link UnsupervisedLabelFormat} configuration:</b></u><br>
+ * Only relevant when the task is set to {@link Task#UNSUPERVISED}. Determine the format of the labels:<br>
+ * <b>RANK2_IDX</b>: return int32 [minibatch, numTokens] array with entries being class numbers<br>
+ * <b>RANK3_NCL</b>: return float32 [minibatch, numClasses, numTokens] array with 1-hot entries along dimension 1<br>
+ * <b>RANK3_NLC</b>: return float32 [minibatch, numTokens, numClasses] array with 1-hot entries along dimension 2<br>
+ * <br>
+ */
+public class BertIterator implements MultiDataSetIterator {
+
+    public enum Task {UNSUPERVISED, SEQ_CLASSIFICATION}
```

Comment:
If I understand correctly, in the paper they are actually pretraining the model with both tasks at the same time. So we should support that case as well.